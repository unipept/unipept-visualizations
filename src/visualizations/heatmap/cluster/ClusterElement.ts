export default class ClusterElement {
    public readonly values: number[];
    public readonly id: number;

    constructor(values: number[], id: number) {
        this.values = values;
        this.id = id;
    }
}
