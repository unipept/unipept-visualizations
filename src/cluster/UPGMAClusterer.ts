import Clusterer from "./clusterer";
import TreeNode from "./treeNode";
import Metric from "../metric/metric";
import ClusterElement from "./clusterElement";
import Cluster from "./cluster";

export default class UPGMAClusterer<T> implements Clusterer<T> {
    private readonly metric: Metric;

    /**
     * @param metric A distance metric that's used for the clustering performed by this class.
     */
    constructor(metric: Metric) {
        this.metric = metric;
    }

    /**
     * This function returns the root of a dendrogram, based upon the given dataset. The clustering is performed on
     * a distance matrix, which is calculated using the metric, defined in the constructor of this class.
     *
     * @param data A matrix containing data elements that should be clustered. The elements are either clustered on row
     *        or column similarity.
     * @param axis The axis against which the clustering should be performed. When given "columns", the elements to be
     *        clustered are considered to be the columns, otherwise the rows are seen as the elements-to-be-clustered.
     */
    cluster(data: ClusterElement<T>[][], axis: "columns" | "rows"): TreeNode<T[]> {
        // When the given axis is columns, the input data needs to be transposed. The rest of the following algorithm
        // works identically for columns or rows.
        if (axis == "columns") {
            let newData: ClusterElement<T>[][] = [];

            // Transpose input array!
            for (let i = 0; i < data[0].length; i++) {
                let row: ClusterElement<T>[] = [];
                for (let j = 0; j < data.length; j++) {
                    row.push(data[j][i]);
                }
                newData.push(row);
            }

            data = newData;
        }

        // A map containing all clusters that are produced during the algorithm.
        let clusters: Map<number, Cluster<T[]>> = new Map();

        // Construct a matrix only containing the values from the given input matrix.
        let valueMatrix: number[][] = [];
        // First we compute the distance matrix according to the given distance metric.
        for (let i = 0; i < data.length; i++) {
            let row: number[] = [];
            for (let j = 0; j < data[i].length; j++) {
                row.push(data[i][j].value);
            }
            valueMatrix.push(row);
            let mappedData: T[] = data[i].map((val) => val.extra);
            // Every item starts in a singleton cluster initially.
            clusters.set(i, new Cluster<T[]>([mappedData], i, new TreeNode<T[]>(null, null, [mappedData], 0)));
        }

        // Effectively compute the distance matrix.
        let distanceMatrix = this.metric.getDistance(valueMatrix);
        let done = 0;

        // Then we keep adding the 2 closest clusters together until there is only one cluster left. During this process
        // a dendrogram is built. This dendrogram is what is returned eventually by this method.
        while (done != distanceMatrix.length - 1) {
            // Compute the distance between all pairs of nodes
            let smallestDistance = Infinity;
            let x = -1;
            let y = -1;
            for (let i of clusters.keys()) {
                for (let j of clusters.keys()) {
                    if (i > j) {
                        if (distanceMatrix[i][j] < smallestDistance) {
                            smallestDistance = distanceMatrix[i][j];
                            x = i;
                            y = j;
                        }
                    }
                }
            }

            // Get the cluster objects corresponding to the closest distance found.
            let xCluster = clusters.get(x);
            let yCluster = clusters.get(y);

            let nodeHeight: number = smallestDistance / 2;

            if (!xCluster || !yCluster) {
                throw "At least one cluster is invalid!";
            }

            // Recalculate distance from this cluster to other clusters (Use average distance)
            // TODO optimize this here! A copy is not necessary here
            let updatedDistanceMatrix: number[][] = this.copyDistanceMatrix(distanceMatrix);
            for (let j of clusters.keys()) {
                if (j != x && j != y) {
                    // Our matrix is lower triangular (because it is symmetric). This means we should extract the value
                    // from the right part of the matrix.
                    let xDistance;
                    if (j > x) {
                        xDistance = distanceMatrix[j][x]
                    } else {
                        xDistance = distanceMatrix[x][j]
                    }

                    let yDistance;
                    if (j > y) {
                        yDistance = distanceMatrix[j][y]
                    } else {
                        yDistance = distanceMatrix[y][j]
                    }

                    // Recalculate the distance between the new, merged cluster and all other clusters.
                    let temp = (xCluster.elements.length * xDistance + yCluster.elements.length * yDistance) / (xCluster.elements.length + yCluster.elements.length);
                    if (j > x) {
                        updatedDistanceMatrix[j][x] = temp;
                    } else {
                        updatedDistanceMatrix[x][j] = temp;
                    }
                }
            }

            distanceMatrix = updatedDistanceMatrix;

            // Merge both new clusters. The height of the TreeNode in the dendrogram associated with this merger, is
            // equal to the distance between clusterY and clusterX in the distanceMatrix, divided by 2.
            xCluster.merge(yCluster, nodeHeight);
            clusters.delete(y);
            ++done;
        }

        return clusters.values().next().value.treeNode;
    }

    private copyDistanceMatrix(distanceMatrix: number[][]): number[][] {
        let output: number[][] = [];

        for (let i = 0; i < distanceMatrix.length; i++) {
            let current: number[] = [];
            let row = distanceMatrix[i];
            for (let j = 0; j < row.length; j++) {
                current.push(row[j]);
            }
            output.push(current);
        }

        return output;
    }
}
