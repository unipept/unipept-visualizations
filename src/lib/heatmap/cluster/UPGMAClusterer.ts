import { copy } from "../../math";
import { Node } from "../../node";
import { Metric } from "../metric/metric";

import { Cluster } from "./cluster";
import { Clusterer } from "./clusterer";


export class UPGMAClusterer implements Clusterer {
  public readonly metric: Metric;

  /**
   * @param metric A distance metric that's used for the clustering performed by this class.
   */
  public constructor(metric: Metric) {
    this.metric = metric;
  }

  /**
   * This function returns the root of a dendrogram, based upon the given dataset.
   * The clustering is performed on a distance matrix, which is calculated using the
   * metric, defined in the constructor of this class.
   *
   * @param data A matrix containing data elements that should be clustered.
   *            The elements are either clustered on row or column similarity.
   */
  public cluster(data: number[][]): Node {
    if (data.length === 0) {
      return new Node({});
    }

    // All clusters that exist in a current step.
    const clusters: Map<number, Cluster>
      = new Map(data.map((v: number[], i: number) => [i, new Cluster(new Node({data: v}), i)]));


    // Now we need to compute a distance matrix.
    // A distance matrix needs a matrix with raw values to be calculated.
    // We thus need to convert the input into a value matrix, before proceeding.

    // Compute the distance matrix!
    let distanceMatrix: number[][] = this.metric.getDistance(data);

    // Start the UPGMA iterations. Loop until only 1 cluster remains.
    let done: number = 0;
    while (done !== distanceMatrix.length - 1) {
      // Look for the smallest value in the distance matrix.
      let smallestDistance: number = Infinity;
      let x: number = -1;
      let y: number = -1;
      for (const i of clusters.keys()) {
        for (const j of clusters.keys()) {
          if (i > j) {
            if (distanceMatrix[i][j] < smallestDistance) {
              smallestDistance = distanceMatrix[i][j];
              x = i;
              y = j;
            }
          }
        }
      }

      // Get the cluster objects corresponding to the closest distance found.
      const xCluster: Cluster | undefined = clusters.get(x);
      const yCluster: Cluster | undefined = clusters.get(y);

      const nodeHeight: number = smallestDistance / 2;

      if (!xCluster || !yCluster) {
        throw new Error("At least one cluster is invalid!");
      }

      // Recalculate distance from this cluster to other clusters (Use average distance)
      const updatedDistanceMatrix: number[][] = copy(distanceMatrix);
      // Cluster.keys() returns a reference to every cluster at the current step
      for (const j of clusters.keys()) {
        if (j !== x && j !== y) {
          // Our matrix is lower triangular (because it is symmetric).
          // This means we should extract the value from the right part of the matrix.
          const xDistance: number = (j > x ? distanceMatrix[j][x] : distanceMatrix[x][j]);
          const yDistance: number = (j > y ? distanceMatrix[j][y] : distanceMatrix[y][j]);

          // Recalculate the distance between the new, merged cluster and all other clusters.
          const temp: number = (xCluster.size() * xDistance + yCluster.size() * yDistance)
            / (xCluster.size() + yCluster.size());
          if (j > x) {
            updatedDistanceMatrix[j][x] = temp;
          } else {
            updatedDistanceMatrix[x][j] = temp;
          }
        }
      }

      distanceMatrix = updatedDistanceMatrix;

      // Merge both new clusters.
      // The height of the tree in the dendrogram associated with this merger, is
      // Equal to the distance between clusterY and clusterX in the distanceMatrix, divided by 2.
      xCluster.merge(yCluster, nodeHeight);
      clusters.delete(y);
      done += 1;
    }

    return clusters.values()
      .next().value.treeNode;
  }
}
