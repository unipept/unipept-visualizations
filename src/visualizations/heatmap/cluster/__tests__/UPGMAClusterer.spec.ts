import UPGMAClusterer from "./../UPGMAClusterer";
import EuclidianDistanceMetric from "./../../metric/EuclidianDistanceMetric";
import ClusterElement from "./../ClusterElement";
import TreeNode from "./../TreeNode";

describe("UPGMAClusterer", () => {
    it("should cluster the provided data as expected", () => {
        const upgmaClusterer = new UPGMAClusterer(new EuclidianDistanceMetric());

        const elements = [
            new ClusterElement([1.0, 2.0, 3.0, 4.0, 5.0], 1),
            new ClusterElement([1.1, 2.1, 3.1, 4.1, 5.1], 2),
            new ClusterElement([7, 8, 9, 10, 11], 3),
            new ClusterElement([3, 4, 5, 6, 7], 4),
            new ClusterElement([3.5, 4.5, 5.5, 6.5, 7.5], 5)
        ];

        // For the data provided above, we expect the following clustering to be performed:
        // 1 and 2 are closest together, so they should be clustered first
        // 4 and 5 are then the two closest items, so they form a second cluster
        // Then {4, 5} will be merged with {1, 2}
        // Finally {1, 2, 4, 5}  will be merged with 3

        const clusters = upgmaClusterer.cluster(elements);

        type PrintableNode = {
            leftChild: PrintableNode | null,
            rightChild: PrintableNode | null,
            id: number[],
        }

        // Helper function that extracts the children and id for each node. This will help in converting the nodes to
        // a printable format that's used for the Jest snapshot.
        function mapToIds(root: TreeNode): PrintableNode {
            const output: PrintableNode = {
                leftChild: null,
                rightChild: null,
                id: []
            };

            if (root.leftChild) {
                output.leftChild = mapToIds(root.leftChild);
            }

            if (root.rightChild) {
                output.rightChild = mapToIds(root.rightChild);
            }

            output.id = root.values.map(v => v.id);

            return output;
        }

        expect(mapToIds(clusters)).toMatchSnapshot();
    });
});
