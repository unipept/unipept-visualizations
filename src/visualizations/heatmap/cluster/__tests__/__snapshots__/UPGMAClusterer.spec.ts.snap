// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UPGMAClusterer > should cluster the provided data as expected 1`] = `
{
  "id": [
    5,
    4,
    2,
    1,
    3,
  ],
  "leftChild": {
    "id": [
      5,
      4,
      2,
      1,
    ],
    "leftChild": {
      "id": [
        5,
        4,
      ],
      "leftChild": {
        "id": [
          5,
        ],
        "leftChild": null,
        "rightChild": null,
      },
      "rightChild": {
        "id": [
          4,
        ],
        "leftChild": null,
        "rightChild": null,
      },
    },
    "rightChild": {
      "id": [
        2,
        1,
      ],
      "leftChild": {
        "id": [
          2,
        ],
        "leftChild": null,
        "rightChild": null,
      },
      "rightChild": {
        "id": [
          1,
        ],
        "leftChild": null,
        "rightChild": null,
      },
    },
  },
  "rightChild": {
    "id": [
      3,
    ],
    "leftChild": null,
    "rightChild": null,
  },
}
`;
